cpp/src/arrow/array.cc

```diff
@@ -492,11 +492,52 @@ DictionaryArray::DictionaryArray(const std::shared_ptr<DataType>& type,
   SetData(data);
 }
 
+Status DictionaryArray::FromArrays(const std::shared_ptr<DataType>& type,
+                                   const std::shared_ptr<Array>& indices,
+                                   std::shared_ptr<Array>* out) {
+  if (indices->length() == 0) {
+    return Status::Invalid("Dictionary indices must have non-zero length");
+  }
+
+  DCHECK_EQ(type->id(), Type::DICTIONARY);
```

Comment:
Since the function documentation says the type is validated, perhaps this should not be a DCHECK, but I'm not sure. @xhochy or @cpcloud have an opinion (same goes for the dcheck below)?